### ISSUE TITLE
Wrong formulation of cost function and derivatives within SODIX

### ISSUE THREAD

**Issue by adku1173** [acoular/acoular#199]
In GitLab by @sebastian.oertwig on Jan 11, 2024, 18:44

It seems that the SODIX cost function and its derivatives in Acoular are different to the formulation presented in the literature. This deviation includes:

1) The cost function in Acoular sums the linear differences and then takes this error meausure to the power of two, whereas the original formulation in the literature considers a squared L2 norm, i.e. the sum of the squared differences.

2) The derivatives of the cost function presented in the literature are only valid for the L2 cost function presented in the literature. Anyway, the derivatives in Acoular differ from those presented in the literature and also from numerically computed derivatives of the cost function in Acoular (see code below). This means, that the derivatives in Acoular do not match any of the cost function formulations (Acoular or literature).

Below is some simple python code that compares the SODIX formulation in Acoular and that from the literature against finite differences. I extracted the few lines from Acoular, split the calculation of the cost function and the gradient into two separate functions, commented the two suspicious lines (one line for the cost function and one for the derivatives), and provided a corrected code within two new functions below.

I also run the SODIX code from Acoular for microphone array data from a static engine noise test and compared the results with our own SODIX. I have clearly seen that the wrong formulation in Acoular leads to spurious artefacts in the calculated source distribution. Unfortunately, I cannot show the comparison of the results here, but I will try to run a simpler testcase with simulated sources from Acoular later.

```python
import numpy as np
import scipy.optimize as opt

# =============================================================================
# PARAMETERS
# =============================================================================
# set the seed to have reproducible results
np.random.seed(123)

# number of sources and microphones
J, M = 30, 20

# random directivities and steering vectors
D = np.random.rand(J,M)
h = np.random.rand(J,M) + 1j*np.random.rand(J,M)
h = h.T

# hermitian CSM
tmp = np.random.uniform(-1, 1, (M, M)) + 1j*np.random.uniform(-1, 1, (M, M))
csm = np.copy(np.tril(tmp) + np.conj(np.tril(tmp, -1).T))
np.fill_diagonal(csm, np.real(csm.diagonal()))


# =============================================================================
# SODIX COST FUNCTION AND GRADIENT FROM ACOULAR
# =============================================================================
def sodix_acoular_cost(D):
    #### the sodix function ####
    Djm = D.reshape([J,M])
    csmmod = np.einsum('jm,jm,jn,jn->mn',h.T,Djm,Djm,h.T.conj() )        
    func = np.sum(np.absolute((csm - csmmod)))**2 # this is different than in the literature
    return func

def sodix_acoular_grad(D):
    ####the sodix  derivitaive ####
    Djm = D.reshape([J,M])
    inner = csm - np.einsum('jl,jl,jm,jm->lm',h.T,Djm,Djm,h.T.conj() )      
    derdrl = -4 *  Djm * np.real(np.einsum('rm,rl,lm->rl',h.T,h.T.conj(),inner)) # here might still be an error
    return derdrl.ravel()


# =============================================================================
# SODIX GRADIENT FROM LITERATURE
# =============================================================================
def sodix_orig_cost(D):
    #### the sodix function ####
    Djm = D.reshape([J,M])
    csmmod = np.einsum('jm,jm,jn,jn->mn',h.T,Djm,Djm,h.T.conj() )        
    func = np.linalg.norm(csm - csmmod)**2
    return func


def sodix_orig_grad(D):
    ####the sodix  derivitaive ####
    Djm = D.reshape([J,M])
    
    inner = csm - np.einsum('jl,jl,jm,jm->lm',h.T,Djm,Djm,h.T.conj() )      
    derdrl = -4 * np.real(np.einsum('rl,rm,ml->rl',h.T,h.T.conj()*Djm,inner))
    return derdrl.ravel()


# =============================================================================
# CHECK GRADIENTS AGAINST FINITE DIFFERENCES
# =============================================================================
err_sodix_acoular_grad = opt.check_grad(
                                func=sodix_acoular_cost,
                                grad=sodix_acoular_grad,
                                x0=D.ravel(),
                                epsilon=np.sqrt(np.finfo(float).eps),
                                direction='all',
                                seed=123,
                                )

err_sodix_orig_grad = opt.check_grad(
                                func=sodix_orig_cost,
                                grad=sodix_orig_grad,
                                x0=D.ravel(),
                                epsilon=np.sqrt(np.finfo(float).eps),
                                direction='all',
                                seed=123,
                                )


print('Normalized error of gradient formulation w.r.t. to cost function in Acoular:',
      err_sodix_acoular_grad / np.linalg.norm(D))
print('Normalized error of gradient formulation w.r.t. to cost function from literature:',
      err_sodix_orig_grad / np.linalg.norm(D))
```

The comparison gives:
```
Normalized error of gradient formulation w.r.t. to cost function in Acoular: 63695.80324945046
Normalized error of gradient formulation w.r.t. to cost function from literature: 0.00013415835156019344
```

**Comment by adku1173** [comment]
In GitLab by @sebastian.oertwig on Jan 14, 2024, 13:51

Thanks to a rainy Sunday morning, I was able to run a simpler testcase that features simulated monopole sound sources from Acoular. In this simulation, there are 8 monopoles aligned in a 2D plane with the same amplitude. The microphone array is a spiral 2D-array with 64 microphones (taken from a publication from Ennes Sarradj) and the source grid has 101 x 101 grid points.

The figures below compare the SODIX result with the correct derivatives (top) and those from Acoular (bottom). I averaged the source directivities derived with SODIX in order to get beamforming-like source maps over the coordinates x and y.

![sim_pos_008_sdx_uniform_noisy_rev000](https://git.tu-berlin.de/acoular-dev/acoular/uploads/02bb83672df98f6d897bda33c0b75792/sim_pos_008_sdx_uniform_noisy_rev000.png)

![sim_pos_008_sdx_uniform_noisy_rev000_Acoular](https://git.tu-berlin.de/acoular-dev/acoular/uploads/9fea456239a67fa98dcf9685f2098c90/sim_pos_008_sdx_uniform_noisy_rev000_Acoular.png)

In total, the source amplitudes are well recovered with both versions,
however, the sources are slightly distributed around their true locations with the formulation in Acoular and there are in general more artefacts.
A reason for both is that the optimization routine stops early when the gradient is not fully correct.
A look into the output of the optimizer in this [line](https://git.tu-berlin.de/acoular-dev/acoular/-/blob/master/acoular/fbeamform.py?ref_type=heads#L2242) gives for this testcase: 

```
dicts = {..., 'task': 'CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH', 'funcalls': 100, 'nit': 15}
``` 

A large deviation between 'funcalls' and 'nit' indicates that the optimization routine had problems to find a new minimum along the provided gradient and therefore, re-computed many function values in a single iteration. In this case, the optimizer hit a different convergence criterium and just stop the optimization.

If the correct derivatives are provided, one will usually find to have similar values for 'funcalls' and 'nit' (in this case 'funcalls'=210, 'nit'=200), which also shows that the optimization continued further with the correct derivatives.
